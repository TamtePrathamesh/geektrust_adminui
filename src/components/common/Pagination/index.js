import React, { useState, useCallback } from 'react';
import PropTypes from 'prop-types';
import LeftArrow from '../../images/left-arrow.png';
import RightArrow from '../../images/right-arrow.png';
import LeadingArrow from '../../images/leading-arrow.png';
import TrailingArrow from '../../images/trailing-arrow.png';

import Style from './style.module.css';

const getPages = (pageCount) => Array(pageCount).fill(0);

const Pagination = ({ pageCount = 0, currentPage = 0, onPageChange = () => null }) => {
  const [cPage, setCurrentPage] = useState(1);

  const nextPage = useCallback(() => {
    if (cPage <= pageCount) {
      let nextPage = currentPage + 1;
      setCurrentPage(nextPage);
      onPageChange(nextPage, pageCount);
    }
  }, [cPage, currentPage, onPageChange, pageCount]);

  const previousPage = useCallback(() => {
    if (cPage >= 1) {
      let previousPage = currentPage - 1;
      setCurrentPage(previousPage);
      onPageChange(previousPage, pageCount);
    }
  }, [cPage, currentPage, onPageChange, pageCount]);

  const pageChange = (page) => {
    if (cPage >= 1 && cPage <= pageCount) {
      setCurrentPage(page);
      onPageChange(page, pageCount);
    }
    if (cPage > pageCount) {
      setCurrentPage(pageCount);
      onPageChange(pageCount, pageCount);
    }
  };

  const firstPage = useCallback(() => {
    setCurrentPage(1);
    onPageChange(1);
  }, [onPageChange]);

  const lastPage = useCallback(() => {
    setCurrentPage(pageCount);
    onPageChange(pageCount);
  }, [onPageChange, pageCount]);

  return (
    <div className={Style.paginationContainer}>
      <div data-testid="leading" className={Style.pagination}>
        <div className={`${Style.actionBtn}`} onClick={firstPage}>
          <img width="10px" height="10px" src={LeadingArrow} alt="cant display" />
        </div>
        <div
          data-testid="previous"
          className={`${Style.actionBtn} ${cPage <= 1 ? Style.disableClick : ''}`}
          onClick={previousPage}
        >
          <img width="10px" height="10px" src={LeftArrow} alt="cant display" />
        </div>
        <div className={Style.contentOverflow}>
          {getPages(pageCount).map((_, index) => (
            <div key={index}>
              <button
                id="stepper"
                className={`${Style.stepper} ${currentPage === index + 1 ? Style.active : ''}`}
                onClick={() => pageChange(index + 1)}
              >
                {index + 1}
              </button>
            </div>
          ))}
        </div>
        <div
          data-testid="next"
          className={`${Style.actionBtn} ${cPage >= pageCount ? Style.disableClick : ''}`}
          onClick={nextPage}
        >
          <img src={RightArrow} alt="cant display" width="10px" height="10px" />
        </div>
        <div data-testid="trailing" className={`${Style.actionBtn}`} onClick={lastPage}>
          <img src={TrailingArrow} alt="cant display" width="10px" height="10px" />
        </div>
      </div>
    </div>
  );
};

Pagination.propTypes = {
  pageCount: PropTypes.number,
  onPageChange: PropTypes.func,
  currentPage: PropTypes.number,
};

export default React.memo(Pagination);
