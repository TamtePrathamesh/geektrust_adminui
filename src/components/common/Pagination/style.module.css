@import '../../../style/colors.css';

.paginationContainer {
  width: fit-content;
}

.pagination {
  width: inherit;
  display: flex;
  justify-content: center;
  align-items: center;
  gap: 15px;
}

.actionBtn {
  cursor: pointer;
  color: var(--black);
  width: 30px;
  height: 30px;
  border-radius: 50%;
  background-color: var(--white);
  border: 1px solid var(--blue_one);
  display: flex;
  justify-content: center;
  align-items: center;
  box-shadow: none;
}

.disableClick {
  pointer-events: none;
  opacity: 0.5;
}

.contentOverflow {
  width: fit-content;
  display: flex;
  gap: 15px;
}

.contentOverflow .active {
  background-color: var(--white);
  border-color: var(--blue_one);
  border: 1px solid var(--blue_one);
  color: var(--blue_one);
  box-shadow: none;
}

.stepper {
  cursor: pointer;
  color: var(--white);
  width: 30px;
  height: 30px;
  border-radius: 50%;
  background-color: var(--blue_one);
  border: 0;
}
