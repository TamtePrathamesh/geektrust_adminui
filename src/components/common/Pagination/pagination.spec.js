import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { screen, render, waitFor } from '@testing-library/react';
import user from '@testing-library/user-event';
import Pagination from '.';

describe('Pagination', () => {
  it('should match snapshot', () => {
    const renderer = new ShallowRenderer();
    expect(renderer.render(<Pagination />)).toMatchSnapshot();
  });

  it('should create no of pages as per pageCount', () => {
    const pageChange = jest.fn();
    expect(render(<Pagination pageCount={2} currentPage={1} onPageChange={pageChange} />).container).toMatchSnapshot();
  });

  it('should disable next and previous button when either at beginning of page or at end of page', () => {
    const pageChange = jest.fn();
    render(<Pagination pageCount={3} currentPage={1} onPageChange={pageChange} />);
    expect(screen.queryByTestId('previous')).toHaveClass('disableClick');
    expect(screen.queryByTestId('next')).not.toHaveClass('disableClick');
    user.click(screen.queryByText('3'));
    expect(screen.queryByTestId('next')).toHaveClass('disableClick');
  });

  it('should take either on first or last page directly when leading and trailing button clicked', async () => {
    const pageChange = jest.fn();
    let currentPage = 2;
    const { rerender } = render(<Pagination pageCount={3} currentPage={currentPage} onPageChange={pageChange} />);
    expect(screen.queryByText(1)).not.toHaveClass('active');
    await user.click(screen.queryByTestId('trailing'));
    currentPage = 1;
    rerender(<Pagination pageCount={3} currentPage={currentPage} onPageChange={pageChange} />);
    await waitFor(() => expect(screen.queryByText(1)).toHaveClass('active'));
  });
});
