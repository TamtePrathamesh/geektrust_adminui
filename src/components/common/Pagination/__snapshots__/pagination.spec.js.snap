// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should create no of pages as per pageCount 1`] = `
<div>
  <div
    class="paginationContainer"
  >
    <div
      class="pagination"
      data-testid="leading"
    >
      <div
        class="actionBtn"
      >
        <img
          alt="cant display"
          height="10px"
          src=""
          width="10px"
        />
      </div>
      <div
        class="actionBtn disableClick"
        data-testid="previous"
      >
        <img
          alt="cant display"
          height="10px"
          src=""
          width="10px"
        />
      </div>
      <div
        class="contentOverflow"
      >
        <div>
          <button
            class="stepper active"
            id="stepper"
          >
            1
          </button>
        </div>
        <div>
          <button
            class="stepper "
            id="stepper"
          >
            2
          </button>
        </div>
      </div>
      <div
        class="actionBtn "
        data-testid="next"
      >
        <img
          alt="cant display"
          height="10px"
          src=""
          width="10px"
        />
      </div>
      <div
        class="actionBtn"
        data-testid="trailing"
      >
        <img
          alt="cant display"
          height="10px"
          src=""
          width="10px"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Pagination should match snapshot 1`] = `
<div
  className="paginationContainer"
>
  <div
    className="pagination"
    data-testid="leading"
  >
    <div
      className="actionBtn"
      onClick={[Function]}
    >
      <img
        alt="cant display"
        height="10px"
        src=""
        width="10px"
      />
    </div>
    <div
      className="actionBtn disableClick"
      data-testid="previous"
      onClick={[Function]}
    >
      <img
        alt="cant display"
        height="10px"
        src=""
        width="10px"
      />
    </div>
    <div
      className="contentOverflow"
    />
    <div
      className="actionBtn disableClick"
      data-testid="next"
      onClick={[Function]}
    >
      <img
        alt="cant display"
        height="10px"
        src=""
        width="10px"
      />
    </div>
    <div
      className="actionBtn"
      data-testid="trailing"
      onClick={[Function]}
    >
      <img
        alt="cant display"
        height="10px"
        src=""
        width="10px"
      />
    </div>
  </div>
</div>
`;
